File: lc_py/lcPy/stack/MinStack.py
class MinStack:
    def __init__(self):
        self.tack = []
        self.min_stack = []
    def push(self,val: int) -> None:
        self.stack.append(val)
        if not self.min_stack:
            min_stack.append(val)
        else:
            self.min_stack.append(min(val,self.min_stack[-1]))      # 访问最后一个元素ww
    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()
    def top(self) -> int:
        return self.stack[-1]
    def getMin(self) -> int:
        return self.min_stack[-1]



if __name__ == "__main__":
    st = MinStack()
    st.push(-2)
    st.push(0)
    st.push(-3)
    print(st.getMin())  # 输出: -3
    st.pop()
    print(st.top())     # 输出: 0
    print(st.getMin())  # 输出: -2

File: lc_java/lcJava/stack/MinStack.java
package stack;

import java.util.Stack;

class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<Integer>();
        minStack = new Stack<Integer>();    
    }
    
    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty()){
            minStack.push(val);
        }else{
            minStack.push(Math.min(val,minStack.peek()));
        }
    }
    
    public void pop() {
        stack.pop();
        minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }

    public static void main(String[] args) {
        MinStack st = new MinStack();
        st.push(-1);
        st.push(2);
        st.pop();
        st.push(13);
        st.push(-22);
        st.push(22);
        System.out.println(st.getMin());
        System.out.println(st.top());
    }
}